Dev/feature 19 global user object

mislavmatijevic opened the pull request:
This PR introduces a lot of code refactoring, making the code much more maintainable.

Previously, many widgets depended on receiving the logged-in user object through their constructors. That got old pretty quickly, but it was hard to fix during rapid app development. The massive size of the first commit of this PR truly shows just how many of these widgets got much cleaner by removing their references to the User class.
Another big issue was that the user object was constantly retrieved from secure storage. Since secure storage is a file (or a cookie on web build), a massive amount of reading of a such data source did the speed of the application no favors.

This PR introduces a lot of enhancements in the APIs as well. It's not unexpected, since APIs are most dependent on the user object. I was thinking about putting the logged-in user in the API class via setter (with no public getter) but was quick to change my mind since the logged-in user is needed in the rest of the app (to show "welcome" messages with the user's first name etc.)

Closes #19.
Before closing of this PR and the connected issue, I would like an opinion.

mislavmatijevic committed:
Implement logged user static field.
This change removes the need to pass a current logged in User object all around the place.
That implements a long-awaited dream.
That was a pun. Get it? Because previously, user was always awaited?
This change also introduces a NewUser class to maintain register functionality. (#19)

mislavmatijevic committed:
Append token and user's username to each request.
Exceptions are login and register routes. (#19)

mislavmatijevic committed:
Improve handling of login and register response.
Response data parsed as UTF-8 string.
Name of the method for setting the current token changed _updateTokenData -> _setToken.

mislavmatijevic committed:
Remember current user data on login.
After every login, a new User object with loaded data is returned.
Login screen takes care of storing the new received user in a global field. (#19)

jfletcher20 reviewed (approved):
### User model changes
I like the changes, definitely a lot cleaner now as you've said yourself.
The whole UserRole thing you have going on with the User model still confuses me, although from what I understand it's that way for the ApiRequests since the API doesn't simply retuern UserRoleType.seller or UserRoleType.buyer, rather an id and a name. Maybe we can look into making some sort of adapter (not sure if "adapter" is the right term for it here) for the UserRole class which would enable us to simply use the enum type UserRoleType and everything else would be automatically parsed for API requests/responses.

Definitely cleaner than you said before, and I can see why it should be better practice than always accessing SecureStorage.

### Ideas for further changes
In preparation for when we redo the backend so that it actually makes sense, I think it might be clever to write a method or class that parses the responsedata, such as "ApiResponse(responseData).status" returning 200 instead of true, or "ApiResponse(responseData).data" returning the data value straight-up, or throwing exceptions if it's null, etc. - perhaps that would be over-engineering, but I think it would make for an easier change to the code later on when we change the way the API works, because all we'd have to do is change how our ApiResponse class and methods return their values.

mislavmatijevic commented:
It wouldn't be over-engineering. It sounds like a good idea, however actual planning of the new backend deserves it's own repository, let alone it's own issue.
I suggest we close this issue as soon as one of us two finds the time to test the entire app with the new way user data is accessed.

mislavmatijevic committed:
Fix crash on register.
Info on newly created user was not being stored in a `loggedIn` user object.
This isn't really the best solution since a user is NOT logged in at this point.
User is not even activated at this point, to be precise - only created. (#19)

mislavmatijevic committed:
Fix assigning store after incomplete registration.
Re-enable user to choose role during first login.
That is, if it was not already done during registration. (#19)

mislavmatijevic committed:
Notify user if store name already in use.

mislavmatijevic committed:
Introduce extension for exception messages.
For some reason, Dart does not support retrieving messages from exceptions.
This solution is from: https://stackoverflow.com/questions/61838741/why-do-i-get-exception-when-i-receive-the-exception-message-in-dart and works just fine.

mislavmatijevic committed:
Refactor user role management.
It is now much easier and safer to work with user's role.

mislavmatijevic committed:
Fix crash on empty store.
Previously, app would encounter an exception if there were no products.

mislavmatijevic committed:
Simplify display of snackbar.
Rewired to use the Message class.

mislavmatijevic committed:
Fix crashing on stores display.
I wasn't very smart when I defined store's balance as int instead of double.

mislavmatijevic committed:
Fix turning login button to loader on login.

mislavmatijevic committed:
Improve login API call.
It automatically stores logged in user and returns nothing.
That way it can intercept if user is registered, but without a role/store. #19

mislavmatijevic committed:
Fix user info not appended to API method.
Adding product to store escaped being refactored when _appendUserToReq method was introduced. (#19)

mislavmatijevic committed:
Improve private API methods arrangement.
Commonly used private methods are on the top.

mislavmatijevic commented:
I added many commits, some are unrelated to user object directly and deal with some general issues with the app. My bad.
